pil-analyzer/src/type_inference.rs

```diff
@@ -649,6 +664,81 @@ impl<'a> TypeChecker<'a> {
         })
     }
 
+    fn statement_type(&self) -> ExpectedType {
+        if self.lambda_kind == FunctionKind::Constr {
+            self.constr_function_statement_type.clone()
+        } else {
+            ExpectedType {
+                ty: Type::Tuple(TupleType { items: vec![] }),
+                allow_array: false,
+            }
+        }
+    }
+
+    fn check_kind_and_expression(
```

Comment:
I think it's probably cleaner to separate this function into two: One for expressions at the end of blocks, and one for isolated expressions (which can be LambaExpressions with their own kind).